### DISCUSSION TITLE
Assigning servos to variables

### DISCUSSION THREAD

**Discussion by freemansoft** [BIPES/BIPES#167] [General]
The output of the `initialize servo` command should be assignable to a variable.

I have two servos connected to my RP2040.    They both get assigned to the variable `servo`.   The "move servo" blocks mandate that the variable be called `servo`

**Comment by jetannenbaum** [comment]
Hi Joe,

   This would be a special use case.  Only the I2C and SPI initialize
blocks work that way.  Sounds like you want to customize BIPES for a
particular need.  Perhaps download from the offline branch and customize
your version?

   Let me know if that works for you.

    Thanks,

      Jet

**Reply by freemansoft** [reply]
I want to control a Sparkfun pan and tilt that has two servos.  BIPES supports a single servo. It initializes it to a hard-coded variable.  The motion controls are hard coded to the same variable name.

UIFlow, again not a great app, supports more than one servo.  It is a bit of a hack actually with a predefined number of servos. I think, again, that it is because they support a limited set of hardware. https://flow.m5stack.com/ .

My code is never anything special. Feel free to ask my code reviewers. :-)

**Comment by jetannenbaum** [comment]
Hi Joe,

   I've read that the Pico (and Pico W) have issues supporting multiple
servos.  The servos draw too much current and burn out the Picos.  I don't
remember what the max number of servis were.  Kevin McAleer talked about it
on one of his SMARs videos.  To that end, a pca9685 servo controller might
be a better option.  I've used one (and at one point ported it to BIPES).
It requires a seperate 5 VDC supply, so as not to burn out the Picos.
Would that work?  Or do you want to stick to trying it on a the Pico W
board?

   Jet

**Reply by freemansoft** [reply]
That is caused by people pulling servo power from the vbus or one of the other power pins.  They should run fine if they power the servos with external power.

Either way, BIPES should support more than one servo. You could do a hard-coded set like UIFlow.  They have a drop list with SErvo1...Servo8.  

I just looked at the code to try and get a feel for how to add a drop list to both of the servo blocks.  There are 25 files that have `init_servo` or `move_servo` in them. They would all have to be modified. Does that sound right?

**Comment by jetannenbaum** [comment]
Hi Joe,You could mimic the PWM functionality. The PWM initialize function assigns a number corresponding to the assigned pin.   Then the PWM frequency and pulse width methods ask for the corresponding number.  You could modify all the init/move functions or just modify the one called by the RP2040JetSent from Jet’s iPhoneOn Jan 15, 2023, at 3:47 PM, Joe Freeman ***@***.***> wrote:﻿
That is caused by people pulling servo power from the vbus or one of the other power pins.  They should run fine if they power the servos with external power.
Either way, BIPES should support more than one servo. You could do a hard-coded set like UIFlow.  They have a drop list with SErvo1...Servo8.
I just looked at the code to try and get a feel for how to add a drop list to both of the servo blocks.  There are 25 files that have init_servo or move_servo in them. They would all have to be modified. Does that sound right?

**Comment by jetannenbaum** [comment]
Hi Joe,

   I added IDs to the servos.  Give it a try and let me know if that works
for you.  I'm still not sure having multiple servos on an RP2040 device is
a good idea; however, I suppose the functionality is there for PWM, so why
not servos.

   Thanks,

     Jet

**Comment by freemansoft** [comment]
I have some code that has the right UI and generates the same code as https://bipes.net.br/ide/  A couple of items came up

1. I'm using Visual Studio Code and it reformatted the whitespace and a couple of multiline breaks for function calls in the .js file I edited.  
    1. What javascript autoformatter are you using if an?  
    2. I could refresh, disable my autoformatter and then add my changes if you are manually formatting
3. My code is broken in the same way as `master` is broken for the rpi [base site](https://bipes.net.br/ide/)   Where is the code for the rpi site [pico site](https://bipes.net.br/pico/ui/)
   1. The pico generated movement function will need to be changed to accept the variable.

**Comment by jetannenbaum** [comment]
Hi Joe,

   I don't maintain the main branch.  It has code that is specific to the
ESP version of MicroPython.

  1.  I'm also using VS Code.  No auto formatter.
  2.  Download from the offline branch.  It is specific for the RP2040
version of MicroPython.  That has my latest, with the ability to have more
than one servo.  Or you can use the online version, also has my updated
multi servo code.  I wanted to be consistent with the PWM code (Init with
an ID tied to the pin).

   Thanks,

    Jet

**Comment by freemansoft** [comment]
That code works great for 2 servos on the PICO RP2040.  Thanks!

![Screenshot 2023-01-15 233907](https://user-images.githubusercontent.com/1217160/212599868-f2fd1dbe-bcf8-47fd-b412-70f69f743930.png)

I can open issues for these.

- You could limit the number of servos in the servo identifier fields by setting a max value.
- The Servo init tooltip may not be right.  It talks about frequencies that may not apply to Servos

The variable name in this code may be misleading.  the duty parameter is a tic count and not an angle.
=== def setServoAngle(servo, angle):
===     angle = -angle
===     if angle < -90:
===             angle = -90
===     elif angle > 90:
===             angle = 90
===     angle = (((angle + 90) / 180) * 6500) + 1500
===     servo.duty_u16(int(angle))

**Comment by jetannenbaum** [comment]
Hi Joe,

   Here is a BIPES hint: use the setting gear to add items:
[image: image.png]
   Much easier than adding multiple create text blocks :-)

 Jet

**Reply by freemansoft** [reply]
The PR I opened fixes the tooltip.

I'll try the settings gear.  Thanks!

**Comment by jetannenbaum** [comment]
Hi Joe,

   Thanks for the PR.  It is cleaner than my original implementation.

    Jet
